pkg/sql/catalog/descs/hydrate.go

```diff
@@ -12,369 +12,165 @@ package descs
-func (tc *Collection) hydrateTypesInDescWithOptions(
-	ctx context.Context,
-	txn *kv.Txn,
-	desc catalog.HydratableDescriptor,
-	includeOffline bool,
-	avoidLeased bool,
-) (catalog.HydratableDescriptor, error) {
-	if desc.Dropped() {
-		return desc, nil
-	}
-	// If there aren't any user defined types in the descriptor, then return
-	// early.
-	if !desc.ContainsUserDefinedTypes() {
-		return desc, nil
-	}
-
-	// It is safe to hydrate directly into Mutable since it is not shared. When
-	// hydrating mutable descriptors, use the mutable access method to access
-	// types.
-	sc, _ := desc.(catalog.SchemaDescriptor)
-	switch t := desc.(type) {
-	case *tabledesc.Mutable:
-		return desc, typedesc.HydrateTypesInTableDescriptor(ctx, t.TableDesc(), getMutableTypeLookupFunc(tc, txn, sc))
-	case *schemadesc.Mutable:
-		return desc, typedesc.HydrateTypesInSchemaDescriptor(ctx, t.SchemaDesc(), getMutableTypeLookupFunc(tc, txn, sc))
-	case *funcdesc.Mutable:
-		return desc, typedesc.HydrateTypesInFunctionDescriptor(ctx, t.FuncDesc(), getMutableTypeLookupFunc(tc, txn, sc))
-	default:
-		// Utilize the cache of hydrated tables if we have one and this descriptor
-		// was leased.
-		// TODO(ajwerner): Consider surfacing the mechanism used to retrieve the
-		// descriptor up to this layer.
-		if tc.canUseHydratedDescriptorCache(desc.GetID()) {
-			hydrated, err := tc.hydrated.GetHydratedDescriptor(
-				ctx, t, getImmutableTypeLookupFunc(tc, txn, includeOffline, avoidLeased, sc),
-			)
-			if err != nil {
-				return nil, err
-			}
-			if hydrated != nil {
-				return hydrated, nil
-			}
+// hydrateDescriptors installs user defined type metadata in all types.T present
+// in the descriptors slice.
+//
+// Hydration is not thread-safe while immutable descriptors are, therefore this
+// method will replace un-hydrated immutable descriptors in the slice with
+// hydrated copies. Mutable descriptors are not thread-safe and so are hydrated
+// in-place. Dropped descriptors do not get hydrated.
+//
+// Optionally, when hydrating we can include offline descriptors and avoid
+// leasing depending on the context. This is set via the flags.
+//
+// Collection method callers expect the descriptors to come back hydrated.
+// In practice, array types here are not hydrated, and that's a bummer.
+// Nobody presently is upset about it, but it's not a good thing.
+// Ideally we'd have a clearer contract regarding hydration and the values
+// stored in the various maps inside the collection. One might want to
+// store only hydrated values in the various maps. This turns out to be
+// somewhat tricky because we'd need to make sure to properly re-hydrate
+// all the relevant descriptors when a type descriptor change. Leased
+// descriptors are at least as tricky, plus, there we have a cache that
+// works relatively well.
+//
+// TODO(ajwerner): Sort out the hydration mess; define clearly what is
+// hydrated where and test the API boundary accordingly.
```

Comment:
Perhaps this TODO can be removed now? I won't claim to have any long-term vision of what hydration should look like.

In any case ` hydrateDescriptors` and friends don't do anything new, it's the old logic but inlined and repackaged a bit.